### ISSUE TITLE
BGR decoded frame has odd coloring 

### ISSUE THREAD

**Issue by GustavoStahl** [NVIDIA/VideoProcessingFramework#139]
When testing and comparing VPF decoding+encoding with OpenCV's, I noticed some different coloring regarding the BGR color space. After isolating the decoders, from both libraries, I was able to generate an image with frames side by side.

_click to enlarge and get more details_
![combine_images](https://user-images.githubusercontent.com/51406079/102662697-1af61c00-415e-11eb-818d-04aa02692d81.png)

Following there's the code used to save those frames, with exception from the "Original", which was taken with a simple PrintScreen.

```python
import cv2
import PyNvCodec as nvc
import numpy as np

video_path = 'video.mp4'

#OpenCV
video = cv2.VideoCapture(video_path)

for _ in range(2):
    success, frame = video.read()
cv2.imwrite('frameOpenCV.png', frame)

#VPF
width, height = 1920, 1080

nvDec = nvc.PyNvDecoder(video_path, 0) 
nvDwn = nvc.PySurfaceDownloader(width, height, nvc.PixelFormat.BGR, 0)
cvtNV12_BGR = nvc.PySurfaceConverter(width, height, nvc.PixelFormat.NV12, nvc.PixelFormat.BGR, 0)

frameBGR = np.ndarray(shape=( nvDec.Framesize() ), dtype=np.uint8)

for _ in range(2):
    success = nvDwn.DownloadSingleSurface(cvtNV12_BGR.Execute(nvDec.DecodeSingleSurface()),
                                                   frameBGR)
                                                   
cv2.imwrite('frameVPF.png', frameBGR.reshape((1080,1920,3)))
```
Any idea how I can fix this?

**Comment by rarzumanyan** [comment]
Hi @GustavoStahl 

>Any idea how I can fix this?

I'm not sure what exactly do you want to fix.
Code sample you've provided consists of 2 steps:

- Decoding which outputs NV12 or YUV420 frames.

- NV12 or YUV420 to BGR conversion.

First step is defined by H.264 / H.265 standards, all compliant decoders produce bit-to-bit same output. So both OpenCV and Nvdec shall provide you with frames which are exactly same.

For the second step (NV12 to BGR color conversion), `nppiNV12ToBGR_8u_P2C3R_Ctx` function is used which may transform YUV420 values to BGR not the same way OpenCV color conversion does.

**Comment by GustavoStahl** [comment]
@rarzumanyan Thanks for the answer, I was curious about this color difference. Eventually my coloring problem was found, which is not the one above and has nothing to do with VPF. Just for curiosity, is there any reason why **RGB➙NV12** and **BGR➙RGB** aren't available? Right now i'm using OpenCV to convert **BGR➙RGB**, Numpy is also possible, and VPF to convert **RGB➙YUV➙NV12**. Thanks for the good work on the library :)

**Comment by rarzumanyan** [comment]
Hi @GustavoStahl 

>why RGB➙NV12 and BGR➙RGB aren't available?

VPF relies on NPP library for GPU-accelerated color conversion.
To my best knowledge, both of these conversions aren't supported.
Here's the link to list of supported conversions: https://docs.nvidia.com/cuda/npp/group__image__color__model__conversion.html

**Comment by kingkang313** [comment]
Hi @rarzumanyan ,  
I also  had a  similar problem.  I used NvDecoder to decoder a video( format is mp4),  got NV12 frame, then use the below code to convert NV12 frame to BGR .   
				Npp8u *pSrc[2];
				pSrc[0] = const_cast<Npp8u *>(pPacketData);
				pSrc[1] = pSrc[0] + pitch*height;
				NppiSize oSizeROI;
				oSizeROI.width =  width;
				oSizeROI.height=  height;
				NppStatus npperr = nppiNV12ToBGR_8u_P2C3R(pSrc,pitch,queue_data->frameDataBGR,width*3,oSizeROI);

And for checking the BGR frame, I use opencv  api to encode the bgr frame to jpg picture. 
1. load BGR to Mat
image = cv::Mat(_height,_width,CV_8UC3 , frame );
2.  save the jpg picture
		imwrite(filename,image);

I save every frame of the video, and save many times. Then compare each picture , I found that about 10 percent of the images had different pixels.  Is this the nppiNV12ToBGR_8u_P2C3R api convert  problem or other? 



![test1](https://user-images.githubusercontent.com/10057978/119799275-806ce680-bf0e-11eb-9528-d05cf082ecd4.png)



![test2](https://user-images.githubusercontent.com/10057978/119799329-8b277b80-bf0e-11eb-84d5-0f85319f6aa8.png)
